library/java/src/io/envoyproxy/envoymobile/engine/AndroidEngineImpl.java

```diff
@@ -2,8 +2,9 @@
 
 import android.content.Context;
 import android.net.ConnectivityManager;
+import io.envoyproxy.envoymobile.engine.types.EnvoyObserver;
 
-public class AndroidEngineImpl {
+public class AndroidEngineImpl implements EnvoyEngine {
```

Comment:
I think this will simplified by subclassing. In particular, I think you want the both the Java and Android versions to share the same `synchronized` loader state, which will only be possible via subclassing.

I also think this really isn't covering the same set of concerns as `EnvoyEngine` over in iOS. In fact, I'm having trouble following the logic here now.

This class was originally designed to be instantiated once only (IIRC), and have that instantiation guarded in a synchronized fashion, to ensure the native DLL is only loaded and initialized once. I'd suggest moving this functionality to an `AndroidEngineLoader` that subclasses an `EnvoyEngineLoader` and then simply having a single `EnvoyEngine(Impl)`.

Reply:
Paired and we went with this: https://github.com/lyft/envoy-mobile/pull/340/commits/c6228343270117789894a5ec01c9de5572c99c56